### PR TITLE
Ramp up

### PR DESCRIPTION
**Pull request by chentex**
# Description

Adding ramp up functionality to the suite.

Each individual test can be provided with the following parameters:

- duration: in minutes
- ramp-type: can be exponential or linear
- start-rate: as in req per second
- end-rate: as in req per second
- ramp-steps: Number of steps to reach the desired end-rate

The test will run # <steps> with a running time of <duration>/<steps> rounded, as each step finishes it will increase the rate according to a delta that is calculated with the parameters:

## For a linear ramp it will use this formula

`delta = ( end-rate - start-rate ) / ( steps - 1 )`
>steps, has always have to be greater than 1

So the new rate will be:

`newRate = oldRate + delta`

## For an exponential distribution:

We are using the exponential formula `f(t)=start-rate * <coeff> ^ t`

the `coeff` is calculated with this formula

`coeff = (end-rate / start-rate) ^ (1 / steps)`

So the new rate will be:

`newRate = start-rate * coeff ^ <# of step>`

### Tasks

- [x] Linear ramp
- [x] Exponential ramp
- [x] Implement usage
- [x] Implement flag options
- [x] Documentation
- [x] Automation tasks
  - [x] Cumulative Avg over time graph
  - [x] Request count over time graph

### CODE REVIEW

File: pkg/tests/run.go [hunk 1]
```diff
@@ -1,19 +1,51 @@
...
 	"github.com/spf13/viper"
 	vegeta "github.com/tsenart/vegeta/v12/lib"
 )
 
-func Run(tc types.TestConfiguration) error {
-	logger := tc.Logger
+// Runner prepares config and runs tests
+type Runner struct {
+	connection      *sdk.Connection
+	ctx             context.Context
```
Comment (troy0820) [inline pkg/tests/run.go:22]:
```suggestion
```
As discussed in previous comment.

**Review by troy0820 — CHANGES_REQUESTED** [review]
@chentex Overall this looks really good.  A lot of good work in this PR 👍🏽.  Just a couple of changes requested and we should be good to go.

**Comment by dry923** [comment]
@chentex Not sure if this functionality is in there (still walking through the code) but can it/it might be useful if it could have a ramp up time and a test time. So it can ramp up for 30 minutes but the entire test will run for 60.

**Comment by chentex** [comment]
So @dry923 if I get you right you wan t something like this:

- Endpoint A:
  - duration: 60 min
  - ramp duration: 30 min
  
And this will run a normal test with a normal `rate` or the `start-rate` for 30 min and then ramp up for 30 min, that it's the remaining of the duration.

File: pkg/ramping/exponential.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import "math"
+
+type Exponential struct {
+	startRate   float64
+	endRtae     float64
+	steps       float64
+	currentRate float64
+	currentStep float64
+	delta       float64
```
[the remark this answers is not in the document]
Comment (chentex) [reply pkg/ramping/exponential.go:11]: For the math operations we needed floats, I will change them to int to better reflect the values, and do the conversions in other palce

File: cmd/ocm-load-test.go [hunk 3]
```diff
@@ -43,19 +41,28 @@ var rootCmd = &cobra.Command{
...
+	rootCmd.Flags().Int("start-rate", 0, "Starting request per second rate. (E.g.: 5 would be 5 req/s)")
+	rootCmd.Flags().Int("end-rate", 0, "Ending request per second rate. (E.g.: 5 would be 5 req/s)")
+	rootCmd.Flags().Int("ramp-steps", 0, "Number of stepts to get from start rate to end rate. (Minimum 2 steps)")
+	rootCmd.Flags().Int("ramp-duration", 0, "Duration of ramp in minutes, before normal execution")
+
+	//Required flags
+	rootCmd.Flags().String("ocm-token", "", "OCM Authorization token")
 	// AWS config
-	// If needed to use multiple AWS account, use the config file
+	// If multiple AWS account are needed use use the config file
```
Comment (dry923) [inline cmd/ocm-load-test.go:65]: Nit but you have "use use" and you probably just want 1.

**Review by dry923 — APPROVED** [review]
LGTM! Great work!

**Comment by troy0820** [comment]
LGTM but can we squash these commits?  NICE Work @chentex !!!! 🎉

**Review by troy0820 — APPROVED** [review]
LGTM 👍🏽

**Pull request merged**